container/multi-value-slice/multi_value_slice.go

```diff
@@ -0,0 +1,417 @@
+				break
+			}
+		}
+		if !found {
+			// This is an optimization. If we didn't find an appended item at index i,
+			// then all larger indices don't have an appended item for the object either.
+			break appendedLoop
+		}
+	}
+
+	srcLen, ok := s.cachedLengths[src.Id()]
+	if ok {
+		s.cachedLengths[dst.Id()] = srcLen
+	}
+}
+
+// Value returns all items for the input object.
+func (s *Slice[V, O]) Value(obj O) []V {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	result := make([]V, len(s.sharedItems))
+	for i, item := range s.sharedItems {
+		ind, ok := s.individualItems[uint64(i)]
+		if !ok {
+			result[i] = item
+		} else {
+			found := false
+			for _, v := range ind.Values {
+				_, found = containsId(v.ids, obj.Id())
+				if found {
+					result[i] = v.val
+					break
+				}
+			}
+			if !found {
+				result[i] = item
+			}
+		}
+	}
+
+	for _, item := range s.appendedItems {
+		found := false
+		for _, v := range item.Values {
+			_, found = containsId(v.ids, obj.Id())
+			if found {
+				result = append(result, v.val)
+				break
+			}
+		}
+		if !found {
+			// This is an optimization. If we didn't find an appended item at index i,
+			// then all larger indices don't have an appended item for the object either.
+			return result
+		}
+	}
+
+	return result
+}
+
+// At returns the item at the requested index for the input object.
+// Appended items' indices are always larger than shared/individual items' indices.
+// We first check if the index is within the length of shared items.
+// If it is, then we return an individual value at that index - if it exists - or a shared value otherwise.
+// If the index is beyond the length of shared values, it is an appended item and that's what gets returned.
+func (s *Slice[V, O]) At(obj O, index uint64) (V, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	if index >= uint64(len(s.sharedItems)+len(s.appendedItems)) {
+		var def V
+		return def, fmt.Errorf("index %d out of bounds", index)
+	}
+
+	isOriginal := index < uint64(len(s.sharedItems))
+	if isOriginal {
+		ind, ok := s.individualItems[index]
+		if !ok {
+			return s.sharedItems[index], nil
+		}
+		for _, v := range ind.Values {
+			for _, id := range v.ids {
+				if id == obj.Id() {
+					return v.val, nil
+				}
+			}
+		}
+		return s.sharedItems[index], nil
+	} else {
+		item := s.appendedItems[index-uint64(len(s.sharedItems))]
+		for _, v := range item.Values {
+			for _, id := range v.ids {
+				if id == obj.Id() {
+					return v.val, nil
+				}
+			}
+		}
+		var def V
+		return def, fmt.Errorf("index %d out of bounds", index)
+	}
+}
+
+// UpdateAt updates the item at the required index for the input object to the passed in value.
+func (s *Slice[V, O]) UpdateAt(obj O, index uint64, val V) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if index >= uint64(len(s.sharedItems)+len(s.appendedItems)) {
+		return fmt.Errorf("index %d out of bounds", index)
+	}
+
+	isOriginal := index < uint64(len(s.sharedItems))
+	if isOriginal {
```

Comment:
isOriginal and its negative could be their own separate methods. The `UpdateAt` method is hard to read and on an initial look